flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBuffer.java

```diff
@@ -0,0 +1,453 @@
+		return processMark(watermark);
+	}
+
+	/**
+	 * Add a {@link LatencyMarker} into queue. A new AsyncCollector will be created and returned.
+	 * <p>
+	 * If queue is full, caller will wait here.
+	 *
+	 * @param latencyMarker LatencyMarker
+	 * @return AsyncCollector
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	public AsyncCollector<IN, OUT> addLatencyMarker(LatencyMarker latencyMarker) throws InterruptedException, IOException {
+		return processMark(latencyMarker);
+	}
+
+	private AsyncCollector<IN, OUT> processMark(StreamElement mark) throws InterruptedException, IOException {
+		while (queue.size() >= bufferSize) {
+			// hold the input StreamRecord until it is placed in the buffer
+			extraStreamElement = mark;
+
+			lock.wait();
+		}
+
+		if (error != null) {
+			throw error;
+		}
+
+		AsyncCollector<IN, OUT> collector = new AsyncCollector(this, true);
+
+		queue.put(collector, mark);
+
+		extraStreamElement = null;
+
+		return collector;
+	}
+
+	/**
+	 * Notify the emitter thread and main thread that an AsyncCollector has completed.
+	 *
+	 * @param collector Completed AsyncCollector
+	 */
+	void markCollectorCompleted(AsyncCollector<IN, OUT> collector) {
+		synchronized (lock) {
+			collector.markDone();
+
+			// notify main thread to keep working
+			lock.notifyAll();
+		}
+	}
+
+	/**
+	 * Caller will wait here if buffer is not empty, meaning that not all async i/o tasks have returned yet.
+	 *
+	 * @throws Exception InterruptedException or IOException from AsyncCollector.
+	 */
+	void waitEmpty() throws InterruptedException, IOException {
+		while (queue.size() != 0) {
+			if (error != null) {
+				throw error;
+			}
+
+			lock.wait();
+		}
+	}
+
+	public void startEmitterThread() {
+		this.emitThread.start();
+	}
+
+	public void stopEmitterThread() {
+		emitter.stop();
+
+		emitThread.interrupt();
+	}
+
+	/**
+	 * Get all StreamElements in the AsyncCollector queue.
+	 * <p>
+	 * Emitter Thread can not output records and will wait for a while due to checkpoiting procedure
+	 * holding the checkpoint lock.
+	 *
+	 * @return A List containing StreamElements.
+	 */
+	public List<StreamElement> getStreamElementsInBuffer() {
+		List<StreamElement> ret = new ArrayList<>(queue.size());
+		for (Map.Entry<AsyncCollector<IN, OUT>, StreamElement> entry : queue.entrySet()) {
+			ret.add(entry.getValue());
+		}
+
+		// add the lonely input outside of queue into return set.
+		if (extraStreamElement != null) {
+			ret.add(extraStreamElement);
+		}
+
+		return ret;
+	}
+
+	@VisibleForTesting
+	public Map<AsyncCollector<IN, OUT>, StreamElement> getQueue() {
+		return this.queue;
+	}
+
+	/**
+	 * A working thread to output results from {@link AsyncCollector} to the next operator.
+	 */
+	private class Emitter implements Runnable {
+		private volatile boolean running = true;
+
+		private void output(AsyncCollector collector, StreamElement element) throws Exception {
```

Comment:
No the `AsyncCollector`. These kind of problems can easily be detected by IntelliJ's inspections, for example.